ARROW-9516: [Rust][DataFusion] refactor of column names

jorgecarleitao opened the pull request:
This PR is respective to ARROW-9516.

It:

1. simplifies how we construct and handle columns: all columns are now name-based, not index-based. This simplifies our code base significantly.
2. makes all column naming happen on the logical plan, not physical plan
3. gives more expressive column names to the end schema of a logical plan, particularly to aggregated expressions (e.g. `SUM(a), SUM(b)` instead of `SUM, SUM`.

This is currently a proof of value: all tests pass and stuff, but there is no decision made on whether we should proceed with these changes. More details available at [ARROW-9516](https://issues.apache.org/jira/browse/ARROW-9516).

jorgecarleitao commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -995,16 +1034,16 @@ mod tests {
 
         let plan = LogicalPlanBuilder::scan("default", "test", schema.as_ref(), None)?
             .aggregate(
-                vec![col("state")],
```
This change was necessary because the file's schema is `c1`, `c2`, not other schema. We were able to get away with this because, since the names did not mean anything, we could read a file with a given schema using another schema's field names, as long as the order was correct. I believe that this should not be possible by design, as it introduces situations that are non-trivial to debug.

jorgecarleitao commented on rust/datafusion/src/logicalplan.rs:
```diff
@@ -179,15 +179,92 @@ impl ScalarValue {
+}
+
+/// Returns the datatype of the expression given the input schema
+// note: the physical plan derived from an expression must match the datatype on this function.
```
This was already the case - I am just making it explicit in this comment.

jorgecarleitao commented:
@andygrove fyi

andygrove commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -455,12 +495,11 @@ impl ExecutionContext {
-            }
-            Expr::Column(i) => {
-                Ok(Arc::new(Column::new(*i, &input_schema.field(*i).name())))
+            Expr::Alias(expr, ..) => Ok(self.create_physical_expr(expr, input_schema)?),
```
It looks like we're losing the alias name here?

andygrove replied:
So we just need to fix this so that the physical expression is created using the aliased name.

jorgecarleitao replied:
No, for the same reason: the physical plan does not care about names anymore, and aliases are only about naming.

jorgecarleitao replied:
fyi my comment was written before I read your second comment!

andygrove replied:
Ok, so the schema for these physical operators is based on the logical schema? I guess I just want to be sure that the alias name is preserved in the query results.

A unit test demonstrating this would be good.

jorgecarleitao replied:
(I wrote my other comment because I became uncertain about my claim and wanted to double check ^_^: yes, good to have this triple checked; this is a big change)

Let's see:

1. the schema of the logical plan is built on the column's names as per [this line](https://github.com/apache/arrow/pull/7796/files#diff-be9adb93b0effc41a9672892e1aed3e1R1012), which is then passed to the physical plan as per line of this discussion. So, it should be the case...
2. [this test](https://github.com/apache/arrow/pull/7796/files#diff-8273e76b6910baa123f3a25a967af3b5R1089) IMO demonstrates this: we alias an expression in the logical plan, and the physical's plan schema's field of that column is the alias itself.

andygrove replied:
Thanks. I just pulled your branch and modified one of the integration tests just to be sure, and it looks good. This really helps simplify some things, thanks!

Could you rebase?

jorgecarleitao replied:
:) perfect. I have rebased this against the latest master.

andygrove commented on rust/datafusion/src/execution/physical_plan/expressions.rs:
```diff
@@ -47,81 +47,43 @@ use arrow::compute::kernels::sort::{SortColumn, SortOptions};
 use arrow::datatypes::{DataType, Schema, TimeUnit};
 use arrow::record_batch::RecordBatch;
 
-/// Represents an aliased expression
```
I don't think we can remove alias from the logical plan. I would expect to see `expr AS alias` in the logical plan when I view it.

jorgecarleitao replied:
I assume that you refer to the physical - not logical - plan in your comment, as this file is about the physical plan. The logical plan does have an `Alias` expression in `logicalplan.rs`.

IMO this is the second core idea of this PR: we do not need aliases as physical expressions at all. The proof for me is 

1. this PR removes `fn name()` from the trait `PhysicalExpr` and 
2. the implementation of `Alias` is: for all method `X(...)` other than `X=name`, `return self.expr.X(...)`

i.e. without `name()`, `Alias` is indistinguishable from its `self.expr` (apart from using extra memory for `self.name`) and thus serves no purpose.

Note that we still have that name in the schema of the physical `projection`, so we can still back-propagate a column name to resolve a column's expression at a given projection (e.g. to optimize projections or perform filter push down). So, all the information is still available at the physical plan.

andygrove replied:
Ah, yes, I was confused. It doesn't make sense to have a physical alias expression.

andygrove commented:
This is looking good overall. I think we need to keep the Alias in the logical plan though.

jorgecarleitao commented:
The test `csv_query_group_by_avg_with_projection` is now failing (after rebase).

The query is `SELECT avg(c12), c1 FROM aggregate_test_100 GROUP BY c1` and the error is that the order of the result is result: it is `c1, avg(c12)`, not `c1, avg(c12)`. I am looking into it.

jorgecarleitao commented:
@andygrove , this is IMO now ready. There were two errors that I fixed in two separate commits:

* 3503c0d fixed an error caused by a mistake of mine
* ff3ade9 fixed an error in one of the tests

The first one was related to the need of adding a projection when the order of SELECT is different from [groups] + [aggregates] in a group by. The current code has this and I had removed that part by mistake in this PR. Also, since this was only being tested in the integration tests, I have added a new test on the library to validate the logical plan.

The second issue is IMO an error in one of the tests: it was parsing `SELECT c2, MIN(c12), MAX(c12) FROM aggregate_test_100 GROUP BY c1` as a valid statement and interpreting that statement as `c2` representing `c1`. With this PR, that is identified as an error because `c2` is not a valid column name in this context.

andygrove commented:
@jorgecarleitao Thanks. It looks like you will have to rebase one more time because master was rebased as part of the 1.0.0 release process. I can merge this once that is done.

jorgecarleitao commented:
@andygrove done. Thanks a lot for the effort and help!

andygrove commented:
@jorgecarleitao The parquet version fix is merged, so one more rebase should get the tests passing.

jorgecarleitao commented:
rebased
